src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,114 @@
+            // both for caching purposes and in order to have a baseline comparison when computing the edits.
+            var previousResultId = tokensCache.Tokens.ResultId == null ? 0 : int.Parse(tokensCache.Tokens.ResultId);
+            var updatedTokens = await SemanticTokensHelpers.ComputeSemanticTokensAsync(
+                request.TextDocument, previousResultId, clientName, SolutionProvider,
+                range: null, cancellationToken).ConfigureAwait(false);
+
+            // If any of the following is true, we do not return any edits and instead only return the fully
+            // computed semantic tokens:
+            // - Previous resultId does not match the cached resultId, or either is null
+            // - Previous document's URI does not match the current document's URI, or either is null
+            // - Previous tokens data or updated tokens data is null
+            if (request.PreviousResultId == null || tokensCache.Tokens.ResultId == null ||
+                request.PreviousResultId != tokensCache.Tokens.ResultId ||
+                request.TextDocument == null || tokensCache.Document == null ||
+                request.TextDocument.Uri != tokensCache.Document.Uri ||
+                tokensCache.Tokens.Data == null || updatedTokens.Data == null)
+            {
+                return new SemanticTokensEditsResult(updatedTokens);
+            }
+
+            var edits = ComputeSemanticTokensEdits(previousResultId, tokensCache.Tokens.Data, updatedTokens.Data);
+            return new SemanticTokensEditsResult(updatedTokens, edits);
+        }
+
+        /// <summary>
+        /// Compares two sets of SemanticTokens and returns the edits between them.
+        /// </summary>
+        private static LSP.SemanticTokensEdits ComputeSemanticTokensEdits(
+            int previousResultId,
+            int[] cachedSemanticTokens,
+            int[] updatedSemanticTokens)
+        {
+            using var _ = ArrayBuilder<SemanticTokensEdit>.GetInstance(out var edits);
+            var index = 0;
+
+            // There are three cases where we might need to create an edit:
+            //     Case 1: Both cached and updated tokens have values at an index, but the tokens don't match
+            //     Case 2: Cached tokens set is longer than updated tokens set - need to make deletion
+            //     Case 3: Updated tokens set is longer than cached tokens set - need to make insertion
+
+            while (index < cachedSemanticTokens.Length && index < updatedSemanticTokens.Length)
+            {
+                // Case 1: Both cached and updated tokens have values at index, but the tokens don't match
+                if (cachedSemanticTokens[index] != updatedSemanticTokens[index])
+                {
+                    edits.Add(GenerateEdit(start: index, deleteCount: 1, data: new int[] { updatedSemanticTokens[index] }));
+                }
+
+                index++;
+            }
```

Comment:
this feels pretty inefficient.  a single edit at the start of a file seems like it would lead to N deletion/inserts. which would actually be more costly than just having the whole result produced.

Should we not be using a diff here between the two sequences, to find a minimal edit changing the old values to the new ones?

Reply:
Agree with this, the recommendation is to do a standard diff of the arrays and just return the minimal edits.

I don't recommend re-implementing diff from scratch if you can help it.

I'm not sure if Roslyn has a diff implementation lying about, but some I am aware of are:
- Google's DiffMatchPatch, ported to C#: https://www.nuget.org/packages?q=diffmatchpatch
- Editor's diff algorithm (unfortunately operates only on strings or TextBuffers): http://index/?query=ITextDifferencing
- Razor language server's diff abstraction